Permit the use of the new query language instead of presentation exchange. (#258)
openid/OpenID4VP: openid-4-verifiable-presentations-1_0.md

endpoint with information about its capabilities as defined in (#request_uri_method_post). The Wallet MAY continue with JAR
when it receives `request_uri_method` parameter with the value `post` but does not support this feature.

> The Verifier MAY articulate requirements of the Credential(s) that are requested using `presentation_definition` and `presentation_definition_uri` parameters that contain a Presentation Definition JSON object as defined in Section 5 of [@!DIF.PresentationExchange]. When processing this Presention Definiotn object, wallet implementations MUST process Presentation Definition JSON object and select candidate Verifiable Credential(s) using the evaluation process described in Section 8 of [@!DIF.PresentationExchange] unless implementing only a credential profile that provides rules on how to evaluate and process [@!DIF.PresentationExchange].

Reviewer: ```suggestion
The Verifier MAY articulate requirements of the Credential(s) that are requested using `presentation_definition` and `presentation_definition_uri` parameters that contain a Presentation Definition JSON object as defined in Section 5 of [@!DIF.PresentationExchange]. When processing this Presentation Definition object, wallet implementations MUST process Presentation Definition JSON object and select candidate Verifiable Credential(s) using the evaluation process described in Section 8 of [@!DIF.PresentationExchange] unless implementing only a credential profile that provides rules on how to evaluate and process [@!DIF.PresentationExchange].
```

> The Verifier MAY articulate requirements of the Credential(s) that are requested using `presentation_definition` and `presentation_definition_uri` parameters that contain a Presentation Definition JSON object as defined in Section 5 of [@!DIF.PresentationExchange]. When processing this Presentation Definition object, wallet implementations MUST process Presentation Definition JSON object and select candidate Verifiable Credential(s) using the evaluation process described in Section 8 of [@!DIF.PresentationExchange] unless implementing only a credential profile that provides rules on how to evaluate and process [@!DIF.PresentationExchange].

Reviewer 2: ```suggestion
The requirements of the Credential(s) that are requested using `presentation_definition` and `presentation_definition_uri` parameters are articulated by the Verified using a Presentation Definition JSON object as defined in Section 5 of [@!DIF.PresentationExchange]. When processing this Presentation Definition object, wallet implementations MUST process Presentation Definition JSON object and select candidate Verifiable Credential(s) using the evaluation process described in Section 8 of [@!DIF.PresentationExchange] unless implementing only a credential profile that provides rules on how to evaluate and process [@!DIF.PresentationExchange].
```

I am highly uncomfortable with a `MAY` here.

This specification defines the following new parameters:

`presentation_definition`:
: OPTIONAL. A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. 

`presentation_definition_uri`:

> : OPTIONAL. A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. See (#request_presentation_definition_uri) for more details.

Reviewer 2: ```suggestion
: OPTIONAL. A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. This parameter MUST NOT be present when `presentation_definition` parameter, or a `scope` value representing a Presentation Definition is present. See (#request_presentation_definition_uri) for more details.
```

> : OPTIONAL. A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. This parameter MUST NOT be present when `presentation_definition` parameter, or a `scope` value representing a Presentation Definition is present. See (#request_presentation_definition_uri) for more details.

Reviewer 2: ```suggestion
: OPTIONAL. A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. This parameter MUST NOT be present if any of the following parameters are present: a `presentation_definition` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined by this specification that indicates what credential is being requested. See (#request_presentation_definition_uri) for more details.
```

> 

Reviewer 2: ```suggestion

One of the following parameters MUST be present to indicate what credential is being requested: a `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined by this specification for that purpose.

```

Author: I think the exact text proposed would prohibit us from introducing a new query language in a future specification without being backwards incompatible. 
The current specification draft does not contain "any other parameter by this specification for that purpose". If this text would be the 1.0 version it means that in order to be compliant with 1.0 you have to use one of the three mechanisms described. 
So even if a future version of the mechanism does describe a new parameter, like vp_query, actually using that would mean you would not be compliant with this proposed language. And therefore using vp_query would not be backwards compatible with 1.0.

It is important that there is always a request parameter present, so we could consider adding the following:
"One of the following parameters MUST be present to indicate what credential is being requested: a `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter for that purpose."

Reviewer 2: `any other parameter defined by this specification for that purpose.` is what intends to allow us to introduce a new query language in a future specification without breaking changes.

Reviewer 3: @Sakurann The problem that @martijnharing has identified, and that I agree with, is the wording "by this specification". This means v1.0 of the specification, and only v1.0 when it is the only published specification. Thus implementors will code for 3 alternative parameters only, since v1.0 only defines 3 parameters. When a later specification comes along that specifies a 4th parameter, v1.0 implementations will be incompatible with this and will say "protocol error" when the 4th parameter is encountered (or may crash!), since from its perspective there are only 3 standardised values. What @martijnharing's proposed edit does is ensures that v1.0 implementations will also code up for any unknown parameter (in v1.0) that might be defined in the future and so should gracefully handle this by replying "<future value> query not supported".

Author: David clarified the issue. I'll try to be more precise:
The issue with the proposed text comes only from it not allowing us to add vp_query in 1.1. If that's not the case then there is no issue, but the reasoning is:

Assumption 1: A change is considered a breaking change if it allows an implementation to be compliant with 1.1 while that implementation is not compliant with 1.0
Assumption 2: It's not allowed to make changes to v1.1 that are considered a breaking change.

Since "this specification" means "OpenID4VP v1.0" in the context of v1.0, the sentence "or any other parameter defined by this specification for that purpose" doesn't add any options because the 3 options listed in that paragraph are the only options defined in OpenID4VP v1.0. Therefore an implementation of a RP according to v1.0 must use one of these 3 options.

If we add the vp_query element to v1.1, then for an RP implementation, using vp_query instead of one of the other 3 options is allowed, since "or any other parameter defined by this specification for that purpose" translates to "or vp_query" in the context of v1.1. 

In this example, this means that: 
- v1.0 effectively says that an RP must use one of 3 options and using anything else is not allowed.
- v1.1 effectively says that an RP must use one of 4 options and using anything else in not allowed.

Since an implementation of an RP under v1.1 using the fourth option (vp_query) is compliant with v1.1 but not compliant with v1.0, actually adding vp_query to v1.1 would be considered a breaking change under assumption 1. 
Therefore under assumption 2 adding vp_query to v1.1 would not be allowed.

If assumption 1 or assumption 2 is not correct (and I happily stand corrected) the concern could very well not apply.

Reviewer 4: Discussed on today's call, change to "defined by this specification (or any future versions of this specification)"

Reviewer 2: @martijnharing could you please accept?

Author: Added the language as was discussed in the call to the PR. I do think it is awkward that compliance to this specification now becomes dependent on the publication schedule of future documents. e.g. test tooling for 1.0 can never be finalized since it is always possible that it must be updated when a new version of this specification becomes available.

This specification defines the following new parameters:

`presentation_definition`:
: OPTIONAL. A string containing a Presentation Definition JSON object. See (#request_presentation_definition) for more details. This parameter MUST NOT be present if any of the following parameters are present: a `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined by this specification that indicates what credential is being requested.

`presentation_definition_uri`:
: OPTIONAL. A string containing an HTTPS URL pointing to a resource where a Presentation Definition JSON object can be retrieved. This parameter MUST NOT be present if any of the following parameters are present: a `presentation_definition` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined by this specification that indicates what credential is being requested. See (#request_presentation_definition_uri) for more details.

> One of the following parameters MUST be present to indicate what credential is being requested: a `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined for that purpose by this specification and all its future versions.

Reviewer 5: What would be a "parameter defined for that purpose defined by this specification"?

> One of the following parameters MUST be present to indicate what credential is being requested: a `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter defined for that purpose by this specification and all its future versions.

Reviewer 5: "all its future versions." means that the parameter has to be defined by all future versions. That's certainly not intended.

> # Response {#response}

Reviewer 5: It feels like this whole section is not consistent (i.e., would need more changes) if we want to make VP Token optional.

Reviewer 2: where can i find how the proposal on how the response with a new query language will look like? without that it's very hard for me to make sense of the changes proposed in the vp token section.

Reviewer 4: See https://github.com/openid/OpenID4VP/issues/178 and the vp token examples, e.g.:

<img width="900" alt="Screenshot 2024-09-25 at 09 22 34" src="https://github.com/user-attachments/assets/2bf87d8f-328f-48a5-aef9-097bba398913">

> A VP Token MAY be returned if the corresponding Authorization Request contained a `presentation_definition` parameter, a `presentation_definition_uri` parameter, or a `scope` parameter representing a Presentation Definition (#vp_token_request).

Reviewer 2: ```suggestion
A VP Token MUST be returned when the corresponding Authorization Request contained a `presentation_definition` parameter, a `presentation_definition_uri` parameter, or a `scope` parameter representing a Presentation Definition (#vp_token_request).
```
this is still a MUST. and honestly, re-reading it, the original wording seems right to me - it says return VP Token when the things asking for it are used. @c2bo - what was wrong in your opinion in the original text?

Reviewer: > A VP Token is only returned if the corresponding Authorization Request contained a `presentation_definition` parameter, a `presentation_definition_uri` parameter, or a `scope` parameter representing a Presentation Definition (#vp_token_request).

We will return a VP Token whenever a request containing some kind of query was provided. In terms of extensibility of query language, I would understand the old text as we will not return a VP Token unless there was a Presentation Definition - which is not what we want with the new query language.

Reviewer 4: The new query language proposal in https://github.com/openid/OpenID4VP/pull/266/files still returns a VP token, so I think the change actually needed here is the same one as being refined above, i.e. adding something about "or any other parameter defined by this specification that indicates what credential is being requested".

Reviewer 2: yeah, then `is only` is totally fine.

> : REQUIRED. JSON String or JSON object that MUST contain a single Verifiable Presentation or an array of JSON Strings and JSON objects each of them containing a Verifiable Presentations. Each Verifiable Presentation MUST be represented as a JSON string (that is a Base64url encoded value) or a JSON object depending on a format as defined in Appendix A of [@!OpenID.VCI].  When a single Verifiable Presentation is returned, the array syntax MUST NOT be used.  If Appendix A of [@!OpenID.VCI] defines a rule for encoding the respective Credential format in the Credential Response, this rules MUST also be followed when encoding Credentials of this format in the `vp_token` response parameter. Otherwise, this specification does not require any additional encoding when a Credential format is already represented as a JSON object or a JSON string.

Reviewer 5: Here, vp_token is still required, above it is optional.

Reviewer 2: umm.. in the new query language, vp_token would still be present, just it's structure would change, no? PR #266 seems to be silent on the response?

Reviewer 5: And this definition MUST be relaxed as well when we want to allow a different structure with a new query language.

`presentation_submission`:

> : OPTIONAL. The `presentation_submission` element as defined in [@!DIF.PresentationExchange]. It contains mappings between the requested Verifiable Credentials and where to find them within the returned VP Token. This is expressed via elements in the `descriptor_map` array, known as Input Descriptor Mapping Objects. These objects contain a field called `path`, which, for this specification, MUST have the value `$` (top level root path) when only one Verifiable Presentation is contained in the VP Token, and MUST have the value `$[n]` (indexed path from root) when there are multiple Verifiable Presentations, where `n` is the index to select. Additional parameters can be defined by Credential Formats, see (#alternative_credential_formats) for details.

Reviewer 2: great point - agreed

> : OPTIONAL. The `presentation_submission` element as defined in [@!DIF.PresentationExchange]. It MUST be present if the `presentation_definition` or `presentation_definition_uri` parameters were present in the request, or if a `scope` value representing a presentation_definition was used. It contains mappings between the requested Verifiable Credentials and where to find them within the returned VP Token. This is expressed via elements in the `descriptor_map` array, known as Input Descriptor Mapping Objects. These objects contain a field called `path`, which, for this specification, MUST have the value `$` (top level root path) when only one Verifiable Presentation is contained in the VP Token, and MUST have the value `$[n]` (indexed path from root) when there are multiple Verifiable Presentations, where `n` is the index to select. Additional parameters can be defined by Credential Formats, see (#alternative_credential_formats) for details.

Reviewer 2: ```suggestion
: OPTIONAL. The `presentation_submission` element as defined in [@!DIF.PresentationExchange]. It MUST be present if the `presentation_definition` parameter, `presentation_definition_uri` parameter, a `scope` parameter with a value representing a Presentation Definition, or any other parameter indicating what credential is being requested defined by this specification was used. It contains mappings between the requested Verifiable Credentials and where to find them within the returned VP Token. This is expressed via elements in the `descriptor_map` array, known as Input Descriptor Mapping Objects. These objects contain a field called `path`, which, for this specification, MUST have the value `$` (top level root path) when only one Verifiable Presentation is contained in the VP Token, and MUST have the value `$[n]` (indexed path from root) when there are multiple Verifiable Presentations, where `n` is the index to select. Additional parameters can be defined by Credential Formats, see (#alternative_credential_formats) for details.
```

Other parameters, such as `state` or `code` (from [@!RFC6749]), or `id_token` (from [@!OpenID.Core]), and `iss` (from [@RFC9207]) can be included in the response as defined in the respective specifications. `state` values MUST only contain ASCII URL safe characters (uppercase and lowercase letters, decimal digits, hyphen, period, underscore, and tilde). For the implementation considerations of a `state` parameter, see (#state_management).

> If present, the `presentation_submission` element MUST be included as a separate response parameter alongside the VP token. Clients MUST ignore any `presentation_submission` element included inside a Verifiable Presentation.

Reviewer 2: actually, @javereec your suggestion changes the meaning and is not correct because when `presentation_submission` is used, it has to be included in the VP Token. @martijnharing's original text is correct.

> The `presentation_submission` element MAY be included as a separate response parameter alongside the VP token. Clients MUST ignore any `presentation_submission` element included inside a Verifiable Presentation.

Reviewer 2: ```suggestion
If present, the `presentation_submission` element MUST be included as a separate response parameter alongside the VP token.
```

Reviewer 5: When present where?

Reviewer 2: where it needs to be according to the specification
